src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java

```diff
@@ -0,0 +1,3394 @@
+         * https://www.tandfonline.com/doi/abs/10.1080/00949655.2015.1060234
+         * Also at https://arxiv.org/abs/1403.6870 (26 March 2014).
+         *
+         * Alastair J. Walker.  1977.  An efficient method for generating discrete random
+         * variables with general distributions. ACM Trans. Math. Software 3, 3
+         * (September 1977), 253-256. DOI: https://doi.org/10.1145/355744.355749
+         *
+         */
+        long U1 = rng.nextLong();
+        // Experimentation on a variety of machines indicates that it is overall much faster
+        // to do the following & and < operations on longs rather than first cast U1 to int
+        // (but then we need to cast to int before doing the array indexing operation).
+        long i = U1 & DoubleZigguratTables.exponentialLayerMask;
+        if (i < DoubleZigguratTables.exponentialNumberOfLayers) {
+            // This is the fast path (occurring more than 98% of the time).  Make an early exit.
+            return DoubleZigguratTables.exponentialX[(int)i] * (U1 >>> 1);
+        }
+        // We didn't use the upper part of U1 after all.  We'll be able to use it later.
+
+        for (double extra = 0.0; ; ) {
+            // Use Walker's alias method to sample an (unsigned) integer j from a discrete
+            // probability distribution that includes the tail and all the ziggurat overhangs;
+            // j will be less than DoubleZigguratTables.exponentialNumberOfLayers + 1.
+            long UA = rng.nextLong();
+            int j = (int)UA & DoubleZigguratTables.exponentialAliasMask;
+            if (UA >= DoubleZigguratTables.exponentialAliasThreshold[j]) {
+                j = DoubleZigguratTables.exponentialAliasMap[j] &
+                        DoubleZigguratTables.exponentialSignCorrectionMask;
+            }
+            if (j > 0) {   // Sample overhang j
+                // For the exponential distribution, every overhang is convex.
+                final double[] X = DoubleZigguratTables.exponentialX;
+                final double[] Y = DoubleZigguratTables.exponentialY;
+                for (;; U1 = (rng.nextLong() >>> 1)) {
+                    long U2 = (rng.nextLong() >>> 1);
+                    // Compute the actual x-coordinate of the randomly chosen point.
+                    double x = (X[j] * 0x1.0p63) + ((X[j-1] - X[j]) * (double)U1);
+                    // Does the point lie below the curve?
+                    long Udiff = U2 - U1;
+                    if (Udiff < 0) {
+                        // We picked a point in the upper-right triangle.  None of those can be
+                        // accepted.  So remap the point into the lower-left triangle and try that.
+                        // In effect, we swap U1 and U2, and invert the sign of Udiff.
+                        Udiff = -Udiff;
+                        U2 = U1;
+                        U1 -= Udiff;
```

Comment:
Updated `U1` never used (recommend running the code through a checker e.g. use IntelliJ)

Reply:
I noticed that before. I think it's a symmetry thing - will check with Guy.